File: Templates/Java/src/main/java/ui/ConsoleUI.java
package ui;

import java.util.List;
import java.util.Scanner;

import models.Student;
import services.StudentService;

public class ConsoleUI {
    private final StudentService studentService;
    private final Scanner scanner;

    public ConsoleUI(StudentService studentService) {
        this.studentService = studentService;
        this.scanner = new Scanner(System.in);
    }

    public void start() {
        boolean exit = false;
        while (!exit) {
            System.out.println("\nStudent Management");
            System.out.println("1. Add Student");
            System.out.println("2. View Student");
            System.out.println("3. Update Student");
            System.out.println("4. Delete Student");
            System.out.println("5. List Students");
            System.out.println("0. Exit");
            System.out.print("Choose an option: ");
            String choice = scanner.nextLine();

            try {
                switch (choice) {
                    case "1" -> this.addStudent();
                    case "2" -> this.viewStudent();
                    case "3" -> this.updateStudent();
                    case "4" -> this.deleteStudent();
                    case "5" -> this.listStudents();
                    case "0" -> exit = true;
                    default -> System.out.println("Invalid option.");
                }
            } catch (Exception error) {
                System.err.println("Error: " + error.getMessage());
            }
        }
    }

    private int readStudentIdUserInput() {
        System.out.print("Enter student ID: ");
        String idInput = scanner.nextLine();
        try {
            int id = Integer.parseInt(idInput);
            return id;
        } catch (NumberFormatException e) {
            System.out.println("Invalid ID. Please enter a number.");
            return this.readStudentIdUserInput();
        }
    }

    private void addStudent() {
        System.out.print("First name: ");
        String firstName = scanner.nextLine();
        System.out.print("Last name: ");
        String lastName = scanner.nextLine();
        System.out.print("Grade: ");
        int grade = Integer.parseInt(scanner.nextLine());
        Student student = new Student(-1, firstName, lastName, grade); // ID will be generated
        this.studentService.addStudent(student);
        System.out.println("Student added.");
    }

    private void viewStudent() {
        int id = this.readStudentIdUserInput();
        Student student = this.studentService.getStudent(id);
        if (student != null) {
            System.out.println(student);
        } else {
            System.out.println("Student not found.");
        }
    }

    private void updateStudent() {
        int id = this.readStudentIdUserInput();
        Student student = this.studentService.getStudent(id);
        if (student == null) {
            System.out.println("Student not found.");
            return;
        }
        System.out.print("First name (" + student.getFirstName() + "): ");
        String firstNameInput = scanner.nextLine();
        String firstName = firstNameInput.isEmpty() ? student.getFirstName() : firstNameInput;
        System.out.print("Last name (" + student.getLastName() + "): ");
        String lastNameInput = scanner.nextLine();
        String lastName = lastNameInput.isEmpty() ? student.getLastName() : lastNameInput;
        System.out.print("Grade (" + student.getGrade() + "): ");
        String gradeInput = scanner.nextLine();
        int grade = gradeInput.isEmpty() ? student.getGrade() : Integer.parseInt(gradeInput);
        Student updatedStudent = new Student(id, firstName, lastName, grade);
        this.studentService.updateStudent(updatedStudent);
        System.out.println("Student updated.");
    }

    private void deleteStudent() {
        int id = this.readStudentIdUserInput();
        this.studentService.deleteStudent(id);
        System.out.println("Student deleted.");
    }

    private void listStudents() {
        List<Student> students = this.studentService.listStudents();
        if (students.isEmpty()) {
            System.out.println("No students found.");
        } else {
            for (Student s : students) {
                System.out.println("Id: " + s.getId() + ", First: " + s.getFirstName() + ", Last: " + s.getLastName()
                        + ", Grade: " + s.getGrade());
            }
        }
    }
}

File: Templates/Kotlin/src/main/kotlin/ui/ConsoleUI.kt
package ui

import models.Student
import services.StudentService
import java.util.Scanner

class ConsoleUI(private val studentService: StudentService) {
    private val scanner = Scanner(System.`in`)

    fun start() {
        var exit = false
        while (!exit) {
            println("\nStudent Management")
            println("1. Add Student")
            println("2. View Student")
            println("3. Update Student")
            println("4. Delete Student")
            println("5. List Students")
            println("0. Exit")
            print("Choose an option: ")
            val choice = scanner.nextLine()

            try {
                when (choice) {
                    "1" -> addStudent()
                    "2" -> viewStudent()
                    "3" -> updateStudent()
                    "4" -> deleteStudent()
                    "5" -> listStudents()
                    "0" -> exit = true
                    else -> println("Invalid option.")
                }
            } catch (error: Exception) {
                System.err.println("Error: ${error.message}")
            }
        }
    }

    private fun readStudentIdUserInput(): Int {
        print("Enter student ID: ")
        val idInput = scanner.nextLine()
        return try {
            idInput.toInt()
        } catch (e: NumberFormatException) {
            println("Invalid ID. Please enter a number.")
            readStudentIdUserInput()
        }
    }

    private fun addStudent() {
        print("First name: ")
        val firstName = scanner.nextLine()
        print("Last name: ")
        val lastName = scanner.nextLine()
        print("Grade: ")
        val grade = scanner.nextLine().toInt()
        val student = Student(-1, firstName, lastName, grade) // ID will be generated
        studentService.addStudent(student)
        println("Student added.")
    }

    private fun viewStudent() {
        val id = readStudentIdUserInput()
        val student = studentService.getStudent(id)
        if (student != null) {
            println(student)
        } else {
            println("Student not found.")
        }
    }

    private fun updateStudent() {
        val id = readStudentIdUserInput()
        val student = studentService.getStudent(id)
        if (student == null) {
            println("Student not found.")
            return
        }
        print("First name (${student.firstName}): ")
        val firstNameInput = scanner.nextLine()
        val firstName = if (firstNameInput.isEmpty()) student.firstName else firstNameInput
        print("Last name (${student.lastName}): ")
        val lastNameInput = scanner.nextLine()
        val lastName = if (lastNameInput.isEmpty()) student.lastName else lastNameInput
        print("Grade (${student.grade}): ")
        val gradeInput = scanner.nextLine()
        val grade = if (gradeInput.isEmpty()) student.grade else gradeInput.toInt()
        val updatedStudent = Student(id, firstName, lastName, grade)
        studentService.updateStudent(updatedStudent)
        println("Student updated.")
    }

    private fun deleteStudent() {
        val id = readStudentIdUserInput()
        studentService.deleteStudent(id)
        println("Student deleted.")
    }

    private fun listStudents() {
        val students = studentService.listStudents()
        if (students.isEmpty()) {
            println("No students found.")
        } else {
            for (s in students) {
                println("Id: ${s.id}, First: ${s.firstName}, Last: ${s.lastName}, Grade: ${s.grade}")
            }
        }
    }
}